File: internal/cli/completion.go
// Package cli provides CLI display helpers.
package cli

import (
	"os"
	"slices"
	"sort"
	"strings"

	"mvmctl/internal/infra"
	"mvmctl/internal/lib/crypto"

	"github.com/spf13/cobra"
)

// --- Shell completion functions ---

// completeNetworkNames completes with network names and short IDs.
func completeNetworkNames(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if opRef == nil {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}
	networks, _ := opRef.NetworkListAll(cmd.Context())
	var results []string
	for _, net := range networks {
		if net.Name != "" && strings.HasPrefix(net.Name, toComplete) && !slices.Contains(results, net.Name) {
			results = append(results, net.Name)
		}
		short := crypto.Truncate(net.ID, 6)
		if strings.HasPrefix(short, toComplete) && !slices.Contains(results, short) {
			results = append(results, short)
		}
	}
	return results, cobra.ShellCompDirectiveNoFileComp
}

// completeImageIDs completes with local image types and short IDs.
func completeImageIDs(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if opRef == nil {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}
	images, _, _ := opRef.ImageListAll(cmd.Context(), false, "", false, nil)
	var results []string
	for _, img := range images {
		short := crypto.Truncate(img.ID, 6)
		if strings.HasPrefix(short, toComplete) && !slices.Contains(results, short) {
			results = append(results, short)
		}
		if img.Type != "" && strings.HasPrefix(img.Type, toComplete) && !slices.Contains(results, img.Type) {
			results = append(results, img.Type)
		}
	}
	return results, cobra.ShellCompDirectiveNoFileComp
}

// completeKernelIDs completes with kernel type:version combos and short IDs.
func completeKernelIDs(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if opRef == nil {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}
	kernels, _, _ := opRef.KernelList(cmd.Context(), false, false, nil)
	var results []string
	for _, k := range kernels {
		if k.Type != "" && k.Version != "" {
			combo := k.Type + ":" + k.Version
			if strings.HasPrefix(combo, toComplete) && !slices.Contains(results, combo) {
				results = append(results, combo)
			}
		}
		short := crypto.Truncate(k.ID, 6)
		if strings.HasPrefix(short, toComplete) && !slices.Contains(results, short) {
			results = append(results, short)
		}
	}
	return results, cobra.ShellCompDirectiveNoFileComp
}

// completeBinaryVersions completes with binary names, versions, and short IDs.
func completeBinaryVersions(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if opRef == nil {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}
	binaries, _, _ := opRef.BinaryList(cmd.Context(), false, nil, nil)
	var results []string
	for _, b := range binaries {
		if b.Type != "" && strings.HasPrefix(b.Type, toComplete) && !slices.Contains(results, b.Type) {
			results = append(results, b.Type)
		}
		if b.Version != "" && strings.HasPrefix(b.Version, toComplete) && !slices.Contains(results, b.Version) {
			results = append(results, b.Version)
		}
		short := crypto.Truncate(b.ID, 6)
		if strings.HasPrefix(short, toComplete) && !slices.Contains(results, short) {
			results = append(results, short)
		}
	}
	return results, cobra.ShellCompDirectiveNoFileComp
}

// completeKeyNames completes with key names, fingerprints (with and without SHA256: prefix).
func completeKeyNames(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if opRef == nil {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}
	keys, _ := opRef.KeyListAll(cmd.Context())
	var results []string
	for _, k := range keys {
		if k.Name != "" && strings.HasPrefix(k.Name, toComplete) && !slices.Contains(results, k.Name) {
			results = append(results, k.Name)
		}
		if k.Fingerprint != "" && strings.HasPrefix(k.Fingerprint, toComplete) &&
			!slices.Contains(results, k.Fingerprint) {
			results = append(results, k.Fingerprint)
		}
		if k.Fingerprint != "" {
			bare := strings.TrimPrefix(k.Fingerprint, "SHA256:")
			if bare != k.Fingerprint && strings.HasPrefix(bare, toComplete) && !slices.Contains(results, bare) {
				results = append(results, bare)
			}
		}
	}
	return results, cobra.ShellCompDirectiveNoFileComp
}

// completeVolumeNames completes with volume names and short IDs.
func completeVolumeNames(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if opRef == nil {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}
	volumes := opRef.VolumeListAll(cmd.Context())
	var results []string
	for _, v := range volumes {
		if v.Name != "" && strings.HasPrefix(v.Name, toComplete) && !slices.Contains(results, v.Name) {
			results = append(results, v.Name)
		}
		short := crypto.Truncate(v.ID, 6)
		if strings.HasPrefix(short, toComplete) && !slices.Contains(results, short) {
			results = append(results, short)
		}
	}
	return results, cobra.ShellCompDirectiveNoFileComp
}

// completeCacheResources completes with cache resource types (static list).
func completeCacheResources(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	resources := []string{"vm", "network", "image", "kernel", "binary", "misc"}
	var results []string
	for _, r := range resources {
		if strings.HasPrefix(r, toComplete) {
			results = append(results, r)
		}
	}
	return results, cobra.ShellCompDirectiveNoFileComp
}

// completeConfigGet completes config args positionally: cat on arg0, key on arg1+.
func completeConfigGet(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if len(args) == 0 {
		return listCategories(toComplete)
	}
	// Second arg: complete keys from the category in args[0]
	if catKeys, ok := infra.OverridableDefaults[args[0]]; ok {
		var keys []string
		for k := range catKeys {
			if strings.HasPrefix(k, toComplete) {
				keys = append(keys, k)
			}
		}
		sort.Strings(keys)
		return keys, cobra.ShellCompDirectiveNoFileComp
	}
	return nil, cobra.ShellCompDirectiveNoFileComp
}

// completeConfigSet completes config set args: cat on arg0, key on arg1.
func completeConfigSet(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if len(args) <= 1 {
		return listCategories(toComplete)
	}
	if catKeys, ok := infra.OverridableDefaults[args[0]]; ok {
		var keys []string
		for k := range catKeys {
			if strings.HasPrefix(k, toComplete) {
				keys = append(keys, k)
			}
		}
		sort.Strings(keys)
		return keys, cobra.ShellCompDirectiveNoFileComp
	}
	return nil, cobra.ShellCompDirectiveNoFileComp
}

// listCategories returns matching config categories.
func listCategories(toComplete string) ([]string, cobra.ShellCompDirective) {
	var cats []string
	for cat := range infra.OverridableDefaults {
		if strings.HasPrefix(cat, toComplete) {
			cats = append(cats, cat)
		}
	}
	sort.Strings(cats)
	return cats, cobra.ShellCompDirectiveNoFileComp
}

// completeRemoteImageIDs completes with remote image IDs (via API).
func completeRemoteImageIDs(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if opRef == nil {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}
	// ImageVersion has no ID field, so remote image ID completion always yields zero results.
	_, _, _ = opRef.ImageListAll(cmd.Context(), true, "", false, nil)
	var results []string
	return results, cobra.ShellCompDirectiveNoFileComp
}

// completeVMThenVolume completes positionally: arg0 = VM names, arg1+ = volume names.
func completeVMThenVolume(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if len(args) == 0 {
		return completeVMNames(cmd, args, toComplete)
	}
	return completeVolumeNames(cmd, args, toComplete)
}

// completeVMThenFile completes positionally: arg0 = VM names, arg1+ = native file completion.
func completeVMThenFile(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if len(args) == 0 {
		return completeVMNames(cmd, args, toComplete)
	}
	return nil, cobra.ShellCompDirectiveDefault
}

// completeKeyThenDir completes positionally: arg0 = key names, arg1+ = directory-only completion.
func completeKeyThenDir(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if len(args) == 0 {
		return completeKeyNames(cmd, args, toComplete)
	}
	return nil, cobra.ShellCompDirectiveFilterDirs
}

// completeVolumeThenSize completes positionally: arg0 = volume names, arg1+ = size suggestions.
func completeVolumeThenSize(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if len(args) == 0 {
		return completeVolumeNames(cmd, args, toComplete)
	}
	sizes := []string{"10G", "20G", "50G", "100G", "512M", "1G", "5G"}
	var results []string
	for _, s := range sizes {
		if strings.HasPrefix(s, toComplete) {
			results = append(results, s)
		}
	}
	return results, cobra.ShellCompDirectiveNoFileComp
}

// completeEnvDestroy completes with workflow state IDs or YAML/YML file paths.
func completeEnvDestroy(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if len(args) > 0 {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}
	var results []string

	// 1. Workflow IDs from state directory
	stateDir := infra.GetWorkflowsStateDir()
	if entries, err := os.ReadDir(stateDir); err == nil {
		for _, e := range entries {
			if e.IsDir() && strings.HasPrefix(e.Name(), toComplete) {
				if !slices.Contains(results, e.Name()) {
					results = append(results, e.Name())
				}
			}
		}
	}

	// 2. YAML/YML files in current working directory
	if cwd, err := os.Getwd(); err == nil {
		if files, err := os.ReadDir(cwd); err == nil {
			for _, f := range files {
				if !f.IsDir() {
					name := f.Name()
					if (strings.HasSuffix(name, ".yaml") || strings.HasSuffix(name, ".yml")) &&
						strings.HasPrefix(name, toComplete) && !slices.Contains(results, name) {
						results = append(results, name)
					}
				}
			}
		}
	}

	return results, cobra.ShellCompDirectiveNoFileComp
}

File: legacy/python/src/mvmctl/cli/_completion.py
"""Shell completion helpers for mvmctl CLI commands.

Each function returns ALL identifier types for its entity, filtered by the
`incomplete` prefix the user has already typed.

All data is fetched through the public API layer — CLI never touches core directly.
"""

from __future__ import annotations


def _complete_remote_image_ids(incomplete: str) -> list[str]:
    """Complete with remote image IDs (via API, not direct asset read)."""
    try:
        from mvmctl.api import ImageOperation

        images = ImageOperation.list_all(remote=True)
        results: list[str] = []
        for img in images:
            if (
                hasattr(img, "id")
                and img.id
                and img.id.startswith(incomplete)
                and img.id not in results
            ):
                results.append(img.id)
        return results
    except Exception:
        return []


def _complete_local_image_ids(incomplete: str) -> list[str]:
    """Complete with local image short IDs and OS slugs."""
    try:
        from mvmctl.api import ImageOperation

        images = ImageOperation.list_all(remote=False)
        results: list[str] = []
        for img in images:
            if hasattr(img, "id") and img.id:
                short = img.id[:6] if len(img.id) >= 6 else img.id
                if short.startswith(incomplete) and short not in results:
                    results.append(short)
            if (
                hasattr(img, "type")
                and img.type
                and img.type.startswith(incomplete)
                and img.type not in results
            ):
                results.append(img.type)
        return results
    except Exception:
        return []


def _complete_vm_names(incomplete: str) -> list[str]:
    """Complete with VM names, short IDs, IPv4 addresses, and MAC addresses."""
    try:
        from mvmctl.api import VMInput, VMOperation

        vms = VMOperation.list_all(VMInput())
        results: list[str] = []
        for vm in vms:
            if (
                vm.name
                and vm.name.startswith(incomplete)
                and vm.name not in results
            ):
                results.append(vm.name)
            if vm.id:
                short = vm.id[:6] if len(vm.id) >= 6 else vm.id
                if short.startswith(incomplete) and short not in results:
                    results.append(short)
            if (
                vm.ipv4
                and vm.ipv4.startswith(incomplete)
                and vm.ipv4 not in results
            ):
                results.append(vm.ipv4)
            if (
                vm.mac
                and vm.mac.startswith(incomplete)
                and vm.mac not in results
            ):
                results.append(vm.mac)
        return results
    except Exception:
        return []


def _complete_network_names(incomplete: str) -> list[str]:
    """Complete with network names and short IDs."""
    try:
        from mvmctl.api import NetworkInput, NetworkOperation

        networks = NetworkOperation.list_all(NetworkInput())
        results: list[str] = []
        for net in networks:
            if (
                net.name
                and net.name.startswith(incomplete)
                and net.name not in results
            ):
                results.append(net.name)
            if net.id:
                short = net.id[:6] if len(net.id) >= 6 else net.id
                if short.startswith(incomplete) and short not in results:
                    results.append(short)
        return results
    except Exception:
        return []


def _complete_kernel_ids(incomplete: str) -> list[str]:
    """Complete with kernel type:version combos and short IDs."""
    try:
        from mvmctl.api import KernelOperation

        kernels = KernelOperation.list_all()
        results: list[str] = []
        for k in kernels:
            if k.type and k.version:
                combo = f"{k.type}:{k.version}"
                if combo.startswith(incomplete) and combo not in results:
                    results.append(combo)
            if k.id:
                short = k.id[:6] if len(k.id) >= 6 else k.id
                if short.startswith(incomplete) and short not in results:
                    results.append(short)
        return results
    except Exception:
        return []


def _complete_binary_versions(incomplete: str) -> list[str]:
    """Complete with binary versions and short IDs."""
    try:
        from mvmctl.api import BinaryOperation

        binaries = BinaryOperation.list_all()
        results: list[str] = []
        for b in binaries:
            if (
                b.version
                and b.version.startswith(incomplete)
                and b.version not in results
            ):
                results.append(b.version)
            if b.id:
                short = b.id[:6] if len(b.id) >= 6 else b.id
                if short.startswith(incomplete) and short not in results:
                    results.append(short)
        return results
    except Exception:
        return []


def _complete_key_names(incomplete: str) -> list[str]:
    """Complete with key names and fingerprints (with and without SHA256: prefix)."""
    try:
        from mvmctl.api import KeyInput, KeyOperation

        keys = KeyOperation.list_all(KeyInput())
        results: list[str] = []
        for k in keys:
            if (
                k.name
                and k.name.startswith(incomplete)
                and k.name not in results
            ):
                results.append(k.name)
            if (
                k.fingerprint
                and k.fingerprint.startswith(incomplete)
                and k.fingerprint not in results
            ):
                results.append(k.fingerprint)
            if k.fingerprint:
                bare = k.fingerprint.removeprefix("SHA256:")
                if (
                    bare != k.fingerprint
                    and bare.startswith(incomplete)
                    and bare not in results
                ):
                    results.append(bare)
        return results
    except Exception:
        return []


def _complete_cache_resources(incomplete: str) -> list[str]:
    """Complete with cache resource names."""
    resources = ["vm", "network", "image", "kernel", "binary", "misc"]
    return [r for r in resources if r.startswith(incomplete)]


def _complete_config_categories(incomplete: str) -> list[str]:
    """Complete with config category names."""
    try:
        from mvmctl.constants import OVERRIDABLE_DEFAULTS

        return [c for c in OVERRIDABLE_DEFAULTS if c.startswith(incomplete)]
    except Exception:
        return []


def _complete_config_keys(incomplete: str) -> list[str]:
    """Complete with config key names from all categories."""
    try:
        from mvmctl.constants import OVERRIDABLE_DEFAULTS

        keys: list[str] = []
        for category_keys in OVERRIDABLE_DEFAULTS.values():
            for key in category_keys:
                if key.startswith(incomplete) and key not in keys:
                    keys.append(key)
        return keys
    except Exception:
        return []


def _complete_volume_names(incomplete: str) -> list[str]:
    """Complete with volume names and short IDs."""
    try:
        from mvmctl.api import VolumeInput, VolumeOperation

        volumes = VolumeOperation.list_all(VolumeInput())
        results: list[str] = []
        for v in volumes:
            if (
                v.name
                and v.name.startswith(incomplete)
                and v.name not in results
            ):
                results.append(v.name)
            if v.id:
                short = v.id[:6] if len(v.id) >= 6 else v.id
                if short.startswith(incomplete) and short not in results:
                    results.append(short)
        return results
    except Exception:
        return []
